server/sonar-server/src/main/java/org/sonar/server/computation/filemove/FileMoveDetectionStep.java

```diff
@@ -0,0 +1,464 @@
+    Snapshot baseProjectSnapshot = analysisMetadataHolder.getBaseProjectSnapshot();
+    if (baseProjectSnapshot == null) {
+      LOG.debug("First analysis. Do nothing.");
+      return;
+    }
+
+    Map<String, ComponentDtoWithSnapshotId> dbFilesByKey = getDbFilesByKey(baseProjectSnapshot);
+
+    if (dbFilesByKey.isEmpty()) {
+      LOG.debug("Previous snapshot has no file. Do nothing.");
+      return;
+    }
+
+    Map<String, Component> reportFilesByKey = getReportFilesByKey(this.rootHolder.getRoot());
+
+    if (reportFilesByKey.isEmpty()) {
+      LOG.debug("No files in report. Do nothing.");
+      return;
+    }
+
+    Set<String> addedFileKeys = ImmutableSet.copyOf(Sets.difference(reportFilesByKey.keySet(), dbFilesByKey.keySet()));
+    Set<String> removedFileKeys = ImmutableSet.copyOf(Sets.difference(dbFilesByKey.keySet(), reportFilesByKey.keySet()));
+
+    // can find matches if at least one of the added or removed files groups is empty => abort
+    if (addedFileKeys.isEmpty() || removedFileKeys.isEmpty()) {
+      LOG.debug("Either no files added or no files removed. Do nothing.");
+      return;
+    }
+
+    // retrieve file data from db and report
+    Map<String, File> dbFileSourcesByKey = getDbFileSourcesByKey(dbFilesByKey, removedFileKeys);
+    Map<String, File> reportFileSourcesByKey = getReportFileSourcesByKey(reportFilesByKey, addedFileKeys);
+
+    // compute score matrix
+    ScoreMatrix scoreMatrix = computeScoreMatrix(dbFileSourcesByKey, reportFileSourcesByKey);
+    printIfDebug(scoreMatrix);
+
+    // not a single match with score higher than MIN_REQUIRED_SCORE => abort
+    if (scoreMatrix.maxScore < MIN_REQUIRED_SCORE) {
+      LOG.debug("max score in matrix is less than min required score (%s). Do nothing.", MIN_REQUIRED_SCORE);
+      return;
+    }
+
+    MatchesByScore matchesByScore = MatchesByScore.create(scoreMatrix);
+
+    ElectedMatches electedMatches = electMatches(dbFileSourcesByKey, reportFileSourcesByKey, matchesByScore);
+
+    for (Match validatedMatch : electedMatches) {
+      movedFilesRepository.setOriginalFile(
+        reportFilesByKey.get(validatedMatch.getReportKey()),
+        toOriginalFile(dbFilesByKey.get(validatedMatch.getDbKey())));
+      LOG.info("File move found: " + validatedMatch);
+    }
+  }
+
+  private Map<String, ComponentDtoWithSnapshotId> getDbFilesByKey(Snapshot baseProjectSnapshot) {
+    try (DbSession dbSession = dbClient.openSession(false)) {
+      return from(dbClient.componentDao().selectAllChildren(
+        dbSession,
+        ComponentTreeQuery.builder()
+          .setBaseSnapshot(new SnapshotDto()
+            .setId(baseProjectSnapshot.getId())
+            .setRootId(baseProjectSnapshot.getId()))
+          .setQualifiers(FILE_QUALIFIERS)
+          .setSortFields(SORT_FIELDS)
+          .setPageSize(Integer.MAX_VALUE)
+          .setPage(1)
+          .build()))
+            .uniqueIndex(ComponentDtoFunctions.toKey());
```

Comment:
Optimization: load only the KEE column in order to not have unused data on the wire.